alibi/explainers/anchors/anchor_text.py

```diff
@@ -0,0 +1,523 @@
+    `temperature`, `use_proba`). `autoregressive` method fills the words one at the time. This corresponds to \
+    multiple forward passes through  the language model which is computationally expensive.
+
+    - ``'sample_proba'`` : ``float`` - probability of a word to be masked.
+
+    - ``'top_n'`` : ``int`` - number of similar words to sample for perturbations.
+
+    - ``'temperature'`` : ``float`` - sample weight hyper-parameter if use_proba equals ``True``.
+
+    - ``'use_proba'`` : ``bool`` - whether to sample according to the predicted words distribution. If set to \
+    ``False``, the `top_n` words are sampled uniformly at random.
+
+    - ``'frac_mask_template'`` : ``float`` - fraction from the number of samples of mask templates to be generated. \
+    In each sampling call, will generate `int(frac_mask_templates * num_samples)` masking templates. \
+    Lower fraction corresponds to lower computation time since the batch fed to the language model is smaller. \
+    After the words' distributions is predicted for each mask, a total of `num_samples` will be generated by sampling \
+    evenly from each template. Note that lower fraction might correspond to less diverse sample. A `sample_proba=1` \
+    corresponds to masking each word. For this case only one masking template will be constructed. \
+    A `filling='autoregressive'` will generate `num_samples` masking templates regardless of the value \
+    of `frac_mask_templates`.
+
+    - ``'batch_size_lm'`` : ``int`` - batch size used for the language model forward pass.
+
+    - ``'punctuation'`` : ``str`` - string of punctuation not to be masked.
+
+    - ``'stopwords'`` : ``List[str]`` - list of words not to be masked.
+
+    - ``'sample_punctuation'`` : ``bool`` - whether to sample punctuation to fill the masked words. If ``False``, the \
+    punctuation defined in `punctuation` will not be sampled.
+"""
+
+
+class AnchorText(Explainer):
+    # sampling methods
+    SAMPLING_UNKNOWN = 'unknown'  #: Unknown sampling strategy.
+    SAMPLING_SIMILARITY = 'similarity'  #: Similarity sampling strategy.
+    SAMPLING_LANGUAGE_MODEL = 'language_model'  #: Language model sampling strategy.
+
+    # default params
+    DEFAULTS: Dict[str, Dict] = {
+        SAMPLING_UNKNOWN: DEFAULT_SAMPLING_UNKNOWN,
+        SAMPLING_SIMILARITY: DEFAULT_SAMPLING_SIMILARITY,
+        SAMPLING_LANGUAGE_MODEL: DEFAULT_SAMPLING_LANGUAGE_MODEL,
+    }
+
+    # class of samplers
+    CLASS_SAMPLER = {
+        SAMPLING_UNKNOWN: UnknownSampler,
+        SAMPLING_SIMILARITY: SimilaritySampler,
+        SAMPLING_LANGUAGE_MODEL: LanguageModelSampler
+    }
+
+    def __init__(self,
+                 predictor: Callable[[List[str]], np.ndarray],
+                 sampling_strategy: str = 'unknown',
+                 nlp: Optional['spacy.language.Language'] = None,
+                 language_model: Union[LanguageModel, MissingDependency, None] = None,
```

Comment:
From looking into [this comment](https://github.com/SeldonIO/alibi-detect/pull/469/files#r833311518) as it's similar to this case here. It turns out what I've done isn't quite right. A simplified version of the above is:

```py
from language_model import LanguageModel # <--- LanguageModel is replaced with a MissingDependency class 

# TestLanguageModel is defined to create a mypy error
class TestLanguageModel:
    def __init__(self, *args):
        pass

def test_fn(a: 'LanguageModel'):
    return a

print(test_fn(LanguageModel())) # Should pass mypy
print(test_fn(TestLanguageModel())) # Should fail mypy
```
So in the above the second function call should fail because TestAsset is not an instance of Asset. Weridly it doesn't fail and I can't actually figure out why but regardless the behaviour is wrong. It can be fixed using:

```py
from language_model import LanguageModel 
from typing import Union, TYPE_CHECKING

if TYPE_CHECKING:
    from language_model.classes import LanguageModel as LanguageModelType

def test_fn(a: 'LanguageModelType'):
    return a

print(test_fn(LanguageModel())) # Does pass mypy
print(test_fn(TestLanguageModel())) # Does fail mypy
```

And this gives the correct `mypy` error:

```
test.py:38: error: Argument 1 to "test_fn" has incompatible type "TestLanguageModel"; expected "LanguageModel"
```

Reply:
Could reproduce with the following (`ray` not present):

```python
from alibi.explainers import DistributedAnchorTabular

class OtherAnchorTabular:
    def __init__(self, *args):
        pass

def test_fn(a: 'DistributedAnchorTabular'):
    return a

print(test_fn(DistributedAnchorTabular()))
print(test_fn(OtherAnchorTabular()))
```

Reply:
`reveal_type(DistributedAnchorTabular)` says it's `Any`

Reply:
Ah, it is the return type from import_optional, because `import_optional` could import anything it's hard to restrict the return type. I suppose typing typically assumes that at import you know what the thing is but then after you've written a tun of code and applied it to that thing you are no longer are so sure and typing fixes that. But in how case by conditionally swapping out the import with `MissingDependency` in `import_optional` we no longer know what something is at import. This just means we have to separate the import of the thing from the import of the type which is what this does:

```
if TYPE_CHECKING:
    from language_model.classes import LanguageModel as LanguageModelType
```

I'll document this in detail in the `import_optional` function. My concern is a developer might import something using `optional_dependency` and use it for typing not realising they've just used `Any` instead of the specific thing they thought they imported. Can you think of anyways of making this fail on misuse?

Reply:
Hmm, just adding a note here of a potential idea but passing Classes to typing constructs like `Union` and `Callable` results in attempts to access the `__class__` attribute. So could issue a warning if this is done which might catch developers using this without being aware that it'll default to Any. It doesn't work in all cases, however, and I'm not sure about any side effects?